q2_amrfinderplus/utils.py

```diff
@@ -15,3 +20,269 @@ def run_command(cmd, cwd=None, verbose=True):
+            pass
+        print(
+            colorify(
+                '"amr_proteins" output is empty because no "--i-proteins" input '
+                "was given."
+            )
+        )
+
+    if not organism:
+        with open(amr_all_mutations.path / "empty_amr_all_mutations.tsv", "w"):
+            pass
+        print(
+            colorify(
+                '"amr_all_mutations" output is empty because no "--p-organism" '
+                "parameter was given."
+            )
+        )
+
+
+def _create_sample_dirs(
+    sequences,
+    proteins,
+    organism,
+    amr_annotations,
+    amr_genes,
+    amr_proteins,
+    amr_all_mutations,
+    sample_id,
+):
+    os.makedirs(amr_annotations.path / sample_id, exist_ok=True)
+    if sequences:
+        os.makedirs(amr_genes.path / sample_id, exist_ok=True)
+    if proteins:
+        os.makedirs(amr_proteins.path / sample_id, exist_ok=True)
+    if organism:
+        os.makedirs(amr_all_mutations.path / sample_id, exist_ok=True)
+
+
+def _create_sample_dict(proteins, sequences):
+    if sequences:
+        # For SampleData[MAGs]
+        if isinstance(sequences, MultiMAGSequencesDirFmt):
+            sample_dict = sequences.sample_dict()
+
+        # For SampleData[Contigs]
+        elif isinstance(sequences, ContigSequencesDirFmt):
+            file_dict = sequences.sample_dict()
+            # Create fake sample for sample_dict
+            sample_dict = {"": file_dict}
+
+        # For FeatureData[MAG]
+        elif isinstance(sequences, MAGSequencesDirFmt):
+            file_dict = sequences.feature_dict()
+            # Create fake sample for sample_dict
+            sample_dict = {"": file_dict}
+
+    else:
+        proteins.pathspec = r".+\.(fa|faa|fasta)$"
+
+        # Monkey patch the sample_dict instance method of MultiMAGSequencesDirFmt to
+        # ProteinsDirectoryFormat if it has a sample data dir structure
+        if any(item.is_dir() for item in proteins.path.iterdir()):
+            proteins.sample_dict = MultiMAGSequencesDirFmt.sample_dict.__get__(
+                proteins, ProteinsDirectoryFormat
+            )
+            sample_dict = proteins.sample_dict()
+        # Monkey patch the feature_dict instance method of MAGSequencesDirFmt to
+        # ProteinsDirectoryFormat if it has a feature data dir structure
+        else:
+            proteins.feature_dict = MAGSequencesDirFmt.feature_dict.__get__(
+                proteins, ProteinsDirectoryFormat
+            )
+            file_dict = proteins.feature_dict()
+            # create sample_dict with fake sample
+            sample_dict = {"": file_dict}
+
+    return sample_dict
+
+
+def _get_file_paths(sequences, proteins, loci, id, file_fp, sample_id=""):
+    # If mags is provided
+    if sequences:
+        dna_path = file_fp
+
+        # If proteins are provided, construct the expected protein file path.
+        if proteins:
+            protein_path = proteins.path / sample_id / f"{id}.fasta"
```

Comment:
What do you mean with safe? 
- Most python packages accept pathlib objects the only thing to remember is it has to be a string when a command for subprocess.run is created.
- Pathlib offers much better path manipulation than os.path.
- If you do .path on a dir format it directly gives you a Path object.
- And it makes everything shorter and more readable.

Reply:
My comment is not about pathlib objects - it's about how you construct your proteins_path altogether. Instead of manually inserting slashes, you can just do: `protein_path = os.path.join(proteins.path, sample_id, f"{id}.fasta")` - it's as readable as what you did (and not all that much longer) but has additional benefits of:
- being explicit about a path being constructed here (as indicated by the method name)
- being cross-platform compatible (keep in mind that paths are not constructed in the same way on all OS's)
- making sure that any double-slashes are corrected (for example, if one of the elements of the path had an additional slash at the end already)